plugout_sdl.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * gbsplay is a Gameboy sound player
+ *
+ * 2020 (C) by Christian Garbs <mitch@cgarbs.de>
+ * Licensed under GNU GPL v1 or, at your option, any later version.
+ *
+ * SDL2 sound output plugin
+ */
+
+#include "common.h"
+
+#include <time.h>
+
+#include <SDL2/SDL.h>
+
+#include "plugout.h"
+
+static const int PLAYBACK_MODE = 0;
+static const int NO_CHANGES_ALLOWED = 0;
+static const int UNPAUSE = 0;
+
+static const struct timespec SLEEP_INTERVAL = {
+	.tv_sec = 0,
+	.tv_nsec = 1000
+};
+
+int device;
+
+static long sdl_open(enum plugout_endian endian, long rate)
+{
+	SDL_AudioSpec desired, obtained;
+	
+	if (SDL_Init(SDL_INIT_AUDIO) != 0) {
+		fprintf(stderr, _("Could not init SDL: %s\n"), SDL_GetError());
+		return -1;
+	}
+
+	// TODO: what about endianess?
+	SDL_zero(desired);
+	desired.freq = rate;
+	desired.format = AUDIO_S16;
+	desired.channels = 2;
+	desired.samples = 1024; // 4096;
```

Comment:
@mrehkopf Should we increase this to 2048 or 4096?
It might reduce choppy audio in MINGW while reducing the update frequency of the terminal output.

Reply:
I'm not sure; I want to look at the differences between the dsound and sdl plugin for a bit. Dsound has pretty quick terminal refresh despite it having a buffer of 4410 samples (at 44100Hz).

Reply:
In any case I think it's desirable to have the buffer size depend on the sample rate (4096 doesn't chop at 44.1kHz but does so a lot at 192kHz).

Reply:
I have tried adaptive samples with this quick hack:

```c
/*
 * SDL_AudioSpec.samples must be a power of 2
 * for 44.1kHz rate, 4096 prevents choppy audio
 * for higher frequencies, more samples are needed
 * dividing by 11 gives a good aproximation for 44.1kHz
 */
static long calculate_samples_from_rate(long rate) {
	long samples_target = rate / 11;
	long samples = 1024;

	while (samples < samples_target)
		samples *= 2;

	fprintf(stdout, _("Debug: SDL samples = %d\n"), samples);
	return samples;
}
```

Result on my Linux system:
* with fixed 1024 samples everything is fine regardless of sample rate (tested 11000, 22050, 44100, 48000 and 192000)
* with adaptive samples, everything larger than 4096 gets choppy (44100Hz -> 4096 samples is ok, 48000Hz -> 8192 samples gets choppy)

Explanations: no idea :-)

Reply:
Choppiness with large bufers probably happens because the value returned by SDL_GetQueuedAudioSize only changes in quantities of the sample buffer size in bytes, so when the buffer becomes too large, you'll only be able to react once it changes to 0 :-)